### ISSUE TITLE
zitadel.user.v2.UserService/ListUsers permission check slow

### ISSUE THREAD

**Issue by muhlemmer** [zitadel/zitadel#9014]
When listing users `zitadel.user.v2.UserService/ListUsers` the response is much slower compared to a v1 list users. Tracing information shows a permission check is performed on each user. Eg: does the requesting user have read permissions on the listed user. On our QA environment (EU1) the permission query and check take around 15-20ms each. The response time on the endpoint is linear to the amount of returned users.

![Image](https://github.com/user-attachments/assets/9a9a78db-0da4-4504-b68a-5a5d97cae267)

We should optimize this behavior. The following alternatives might work better:

1. If a permission is granted through the resource owner (organization), we need to query the org only once and cache for all the other entries with the same resource owner. This still might result in multiple queries if users from multiple organizations are returned.
2. Query for a list of organization the requesting users has the proper permissions for once and apply that as a filter. This might be completely resolvable inside the database using a sub-query or CTE.
3. Move all permissions into the database. A [discussion](https://github.com/zitadel/zitadel/discussions/7256) for this is already opened. This solution would definitely take more time and for now it might be better to stick with the other options.

Discussed with @stebenz 
Original report: https://discord.com/channels/927474939156643850/1311303847695618119

**Comment by muhlemmer** [comment]
Discussed with @hifabienne and @adlerhurst that it's probably best if we go for option 3. I've created a sub-issue for that.

**Comment by ashep** [comment]
Hello, @muhlemmer! Can you please tell if there are any updates?

**Comment by muhlemmer** [comment]
Yes, I started work on something before the holiday. But there still lots of things to iron out. One of the biggest problems I'm facing is that CockroachDB doesn't support `RETURN QUERY` for Pl/PgSQL: https://github.com/cockroachdb/cockroach/issues/117744

I'll try to continue on this next week.
